### PR TITLE
Combination flips v2

### PR DESCRIPTION
**Pull request by zumaad**
Most of the logic is similar to that in the v1 pr. The main difference is that recipe flips exist outside of the context of flipping items. To that end, recipe flips have their own tab:

Screen Shot 2022-01-31 at 7 53 58 PM

Within that tab are recipe flip groups, which contain each recipe flip made that followed that recipe flip group's recipe:
Screen Shot 2022-01-31 at 7 55 11 PM

### CODE REVIEW

File: src/main/java/com/flippingutilities/controller/RecipeHandler.java [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+            int maxRecipesThatCanBeMade = (int) (totalQuantity / quantityNeededForRecipe);
+            return Map.entry(itemId, maxRecipesThatCanBeMade);
+        }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    public void addRecipeFlip(List<RecipeFlipGroup> recipeFlipGroups, RecipeFlip recipeFlip) {
+        int someInputItemId = recipeFlip.getInputs().keySet().iterator().next();
+        Optional<Recipe> recipeMaybe = getApplicableRecipe(someInputItemId, true);
+        if (recipeMaybe.isEmpty()) {
+            log.info("not adding recipe, for some reason the recipe flip does not have an associated recipe");
```
Comment (AntonyGarand) [inline src/main/java/com/flippingutilities/controller/RecipeHandler.java:231]: warn?

File: src/main/java/com/flippingutilities/controller/RecipeHandler.java [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+        }
+        int itemId = offers.get(0).getItemId();
+        //get all the rfgs containing recipe flips that the offers could possibly be referenced in
+        recipeFlipGroups.stream().filter(rfg -> rfg.isInGroup(itemId)).forEach(rfg -> rfg.deleteFlipsWithDeletedOffers(offers));
+    }
+
+    private Optional<List<Recipe>> loadRecipes() {
+        Request request = new Request.Builder()
+                .url("https://raw.githubusercontent.com/Flipping-Utilities/osrs-datasets/master/recipes.json")
+                .build();
```
Comment (AntonyGarand) [inline src/main/java/com/flippingutilities/controller/RecipeHandler.java:302]: Should we cache this locally for a day or so?

File: src/main/java/com/flippingutilities/model/RecipeFlip.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                    .map(entry -> Map.entry(entry.getKey(), entry.getValue().clone()))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    public long getProfit() {
+        return getRevenue() - getExpense();
+    }
+
+    private long getExpenseOrRevenue(boolean getExpense) {
```
Comment (AntonyGarand) [inline src/main/java/com/flippingutilities/model/RecipeFlip.java:62]: method name: `getIngredientsValue(isBuyOffer)`?
I would even make an enum with true/false for the clear name, but renaming `getExpense` to `isBuyOffer` should be better already imo

File: src/main/java/com/flippingutilities/model/RecipeFlipGroup.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        allRecipeFlips.addAll(otherRecipeFlipGroup.getRecipeFlips());
+        allRecipeFlips.sort(Comparator.comparing(RecipeFlip::getTimeOfCreation));
+        return new RecipeFlipGroup(recipe, allRecipeFlips);
+    }
+
+    public Instant getLatestActivityTime() {
+        if (recipeFlips.isEmpty()) {
+            return Instant.EPOCH;
+        }
+        return recipeFlips.get(recipeFlips.size()-1).timeOfCreation;
```
Comment (AntonyGarand) [inline src/main/java/com/flippingutilities/model/RecipeFlipGroup.java:37]: Is the last one always the latest?

File: src/main/java/com/flippingutilities/ui/flipping/FlippingPanel.java [hunk 5]
```diff
@@ -205,104 +197,31 @@ public void rebuild(List<FlippingItem> flippingItems)
...
-					float item1roi = (float) item1ProfitEach / item1.getLatestInstaBuy().get().getPrice() * 100;
-					float item2roi = (float) item2ProfitEach / item2.getLatestInstaBuy().get().getPrice() * 100;
-
-					return Float.compare(item1roi, item2roi);
-				});
-				break;
+		if (favoriteSelected && isItemHighlighted()) {
+			result = result.stream().filter(FlippingItem::isFavorite).collect(Collectors.toList());
 		}
+		sortByTime(result); //when it is on favorites you also want it to be sorted by time
```
Comment (AntonyGarand) [inline src/main/java/com/flippingutilities/ui/flipping/FlippingPanel.java:224]: shouldn't this be in `if(favoriteSelected)` then?

**Review by AntonyGarand — APPROVED** [review]
Lgtm!

File: src/main/java/com/flippingutilities/controller/RecipeHandler.java (hunk 2 above)
Comment (zumaad) [reply src/main/java/com/flippingutilities/controller/RecipeHandler.java:302]: I guess we could, but it is only requested at plugin startup, which shouldn't be too bad.

**Pull request merged**